docs/proposals/scheduling/20220701-pod-migration-job.md

```diff
@@ -0,0 +1,397 @@
+
+## Motivation
+
+Migrating Pods is an important capability that many components (such as deschedulers) rely on, and can be used to optimize scheduling or help resolve workload runtime quality issues. We believe that pod migration is a complex process, involving steps such as auditing, resource allocation, and application startup, and is mixed with application upgrading, scaling scenarios, and resource operation and maintenance operations by cluster administrators. Therefore, how to manage the stability risk of this process to ensure that the application does not fail due to the migration of Pods is a very critical issue that must be resolved.
+
+Therefore, it is necessary to realize a final state-oriented migration capability based on CRD, track the status of each process in the migration, and perceive scenarios such as upgrading and scaling of the application.
+
+### Goals
+
+1. Defines a CRD-based Pod Migration Job API, through which the descheduler can evict or delete Pods more safely.
+2. Describe in detail the design details behind the API.
+
+### Non-Goals/Future Work
+
+1. A new descheduler framework
+2. Descheduling capability for different scenarios such as load-aware descheduling, defragemention, etc.
+3. The details about Deterministic preemption that preempts other Pods for Reservation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+The descheduler in the K8s community evicts pods to be rescheduled according to different strategies. However, it does not guarantee whether the evicted Pod has resources available after re-creation. If a large number of new Pods are in the Pending state when the resources in the cluster are tight, may lower the application availabilities.
+
+#### Story 2
+
+The descheduler evicts the Pod through the Eviction API, and the Eviction API decides whether to delete the Pod according to the PDB status. However, it is unable to perceive workload upgrades, scaling and other scenarios in which Pods are deleted, which will also bring security risks.
+
+#### Story 3
+
+The Pod migration capability itself can be provided to users as a service. Users can integrate this API in their own systems to achieve safe migration, and are no longer limited to deschedulers.
+
+
+### Basic Migration API
+
+These APIs provide cluster administrators with more fine-grained migration control capabilities, which can better reduce risks.
+
+- `scheduling.koordinator.sh/eviction-cost` indicates the cost of eviction. Although the K8s community has [Pod Deletion Cost #2255](https://github.com/kubernetes/enhancements/issues/2255), it is not a general mechanism. To avoid conflicts with components that use `Pod Deletion Cost`, users can individually mark the eviction cost for Pods. The higher the value, the later in the eviction order, and the cost is set to `math.MaxInt` when migration is not allowed.
```

Comment:
Another story is how the combination of <priority, eviction-cost > takes effect. For example, whether it is the same priority, eviction-cost is considered？

Reply:
Int is 64bit in golang, the MaxInt number is too large to write. Think about MaxInt32, MaxUint32? Look at the deletion-cost constraint.

Reply:
![image](https://user-images.githubusercontent.com/4395943/178640214-f3819fd0-db7d-4666-91d8-8e0bfb09a2d0.png)

Reply:
done.